diff --git a/go.mod b/go.mod
index b9868c4645..f85e8e2d3f 100644
--- a/go.mod
+++ b/go.mod
@@ -1,6 +1,6 @@
 module dawn.googlesource.com/dawn
 
-go 1.18
+go 1.24
 
 require (
 	cloud.google.com/go/bigquery v1.52.0
diff --git a/package-lock.json b/package-lock.json
new file mode 100644
index 0000000000..0a62c23172
--- /dev/null
+++ b/package-lock.json
@@ -0,0 +1,6 @@
+{
+  "name": "dawn",
+  "lockfileVersion": 3,
+  "requires": true,
+  "packages": {}
+}
diff --git a/tools/src/cmd/run-cts/node/cmd.go b/tools/src/cmd/run-cts/node/cmd.go
index b41c0af5dc..407151a34d 100644
--- a/tools/src/cmd/run-cts/node/cmd.go
+++ b/tools/src/cmd/run-cts/node/cmd.go
@@ -35,6 +35,7 @@ import (
 	"os/exec"
 	"path/filepath"
 	"runtime"
+	//"strings"
 
 	"dawn.googlesource.com/dawn/tools/src/cmd/run-cts/common"
 	"dawn.googlesource.com/dawn/tools/src/cov"
@@ -61,6 +62,7 @@ type flags struct {
 	debugCTS             bool
 	skipVSCodeInfo       bool
 	enforceDefaultLimits bool
+	mutantTracking		 bool
 	dawn                 node.Flags
 }
 
@@ -113,11 +115,13 @@ func (c *cmd) RegisterFlags(ctx context.Context, cfg common.Config) ([]string, e
 	flag.BoolVar(&c.flags.debugCTS, "debug-cts", false, "enable CTS debugging option")
 	flag.BoolVar(&c.flags.skipVSCodeInfo, "skip-vs-code-info", false, "skips emitting VS Code information")
 	flag.BoolVar(&c.flags.enforceDefaultLimits, "enforce-default-limits", false, "enforce the default limits (note: powerPreference tests may fail)")
+	flag.BoolVar(&c.flags.mutantTracking, "mutant-tracking", false, "record covered mutants in tracking file")
 
 	return []string{"[query]"}, nil
 }
 
 func (c *cmd) Run(ctx context.Context, cfg common.Config) error {
+
 	// Process the command line flags
 	if err := c.processFlags(); err != nil {
 		return err
@@ -137,8 +141,8 @@ func (c *cmd) Run(ctx context.Context, cfg common.Config) error {
 		return fmt.Errorf("failed to gather tests: %w", err)
 	}
 	fmt.Printf("Testing %d test cases...\n", len(testCases))
-
-	var runner func(ctx context.Context, testCases []common.TestCase, results chan<- common.Result)
+	
+	var runner func(ctx context.Context, testCases []common.TestCase, results chan<- common.Result, tracking bool)
 	if c.flags.isolated {
 		fmt.Println("Running in parallel isolated...")
 		runner = c.runTestCasesWithCmdline
@@ -149,7 +153,7 @@ func (c *cmd) Run(ctx context.Context, cfg common.Config) error {
 
 	resultStream := make(chan common.Result, 256)
 	go func() {
-		runner(ctx, testCases, resultStream)
+		runner(ctx, testCases, resultStream, c.flags.mutantTracking)
 		close(resultStream)
 	}()
 
diff --git a/tools/src/cmd/run-cts/node/cmdline.go b/tools/src/cmd/run-cts/node/cmdline.go
index ee9b691ac6..3810072658 100644
--- a/tools/src/cmd/run-cts/node/cmdline.go
+++ b/tools/src/cmd/run-cts/node/cmdline.go
@@ -46,7 +46,7 @@ import (
 // test case in a separate process. This reduces possibility of state leakage
 // between tests.
 // Up to c.flags.numRunners tests will be run concurrently.
-func (c *cmd) runTestCasesWithCmdline(ctx context.Context, testCases []common.TestCase, results chan<- common.Result) {
+func (c *cmd) runTestCasesWithCmdline(ctx context.Context, testCases []common.TestCase, results chan<- common.Result, tracking bool) {
 	// Create a chan of test indices.
 	// This will be read by the test runner goroutines.
 	testCaseIndices := make(chan int, 256)
diff --git a/tools/src/cmd/run-cts/node/server.go b/tools/src/cmd/run-cts/node/server.go
index 915ffbb3f9..dae9e68396 100644
--- a/tools/src/cmd/run-cts/node/server.go
+++ b/tools/src/cmd/run-cts/node/server.go
@@ -34,11 +34,13 @@ import (
 	"fmt"
 	"io"
 	"net/http"
+	//"maps"
 	"os"
 	"os/exec"
 	"path/filepath"
 	"strconv"
 	"strings"
+	//"slices"
 	"sync"
 	"time"
 
@@ -46,10 +48,63 @@ import (
 	"dawn.googlesource.com/dawn/tools/src/utils"
 )
 
+func saveMappingToFile(filename string, testmap map[string]string) {
+
+	//TODO: Make this file write cleaner
+
+	f, err := os.Create(filename)
+	if err != nil {
+		panic(err)
+	}
+
+	defer f.Close()
+	n1, err := f.WriteString("{")
+	if err != nil {
+		panic(err)
+	}
+	fmt.Printf("Write to file %d", n1)
+
+	var firstline bool = true
+	for k, v := range testmap {
+
+		line := string("\"" + strings.Replace(string(k),"\"","\\\"",-1) + "\" : \"" + v + "\"")
+		
+		if !firstline {
+			line = string(",\n" + line)
+		}
+
+		n2, err := f.WriteString(line)
+		if err != nil {
+			panic(err)
+		}
+		fmt.Printf("Write to file %d", n2)
+		firstline = false
+	}
+	n3, err := f.WriteString("}")
+	if err != nil {
+		panic(err)
+	}
+	fmt.Printf("Write to file %d", n3)
+
+	f.Sync()
+	
+}
 // runTestCasesWithServers spawns c.flags.NumRunners server instances to run all
 // the test cases in testCases. The results of the tests are streamed to results.
 // Blocks until all the tests have been run.
-func (c *cmd) runTestCasesWithServers(ctx context.Context, testCases []common.TestCase, results chan<- common.Result) {
+func (c *cmd) runTestCasesWithServers(ctx context.Context, testCases []common.TestCase, results chan<- common.Result, tracking bool) {
+
+	testIDMap := make(map[string]string)
+
+	if (tracking) {
+		// Create mapping from test to id (used to name the test file)
+		for i:= 0; i < len(testCases); i++ {
+			testIDMap[string(testCases[i])] = fmt.Sprintf("test_id_%d",i)
+		}
+		// Store test case index mapping in file
+		saveMappingToFile("/data/dev/dredd-webgpu-testing/llvmpipe/output/covered_by_cts/test_wise_tracking/mapping_test_to_id.json", testIDMap)
+	}
+
 	// Create a chan of test indices.
 	// This will be read by the test runner goroutines.
 	testCaseIndices := make(chan int, 256)
@@ -67,12 +122,13 @@ func (c *cmd) runTestCasesWithServers(ctx context.Context, testCases []common.Te
 		wg.Add(1)
 		go func() {
 			defer wg.Done()
-			if err := c.runServer(ctx, id, testCases, testCaseIndices, results); err != nil {
+			if err := c.runServer(ctx, id, testCases, testCaseIndices, results, testIDMap); err != nil {
 				results <- common.Result{
 					Status: common.Fail,
 					Error:  fmt.Errorf("Test server error: %w", err),
 				}
 			}
+
 		}()
 	}
 
@@ -84,12 +140,13 @@ func (c *cmd) runTestCasesWithServers(ctx context.Context, testCases []common.Te
 // The result of the test run is written to the results chan.
 // Once the testCaseIndices chan has been closed and all taken tests have been
 // completed, the server process is shutdown and runServer returns.
-func (c *cmd) runServer(
+func (c *cmd) runServer( 
 	ctx context.Context,
 	id int,
 	testCases []common.TestCase,
 	testCaseIndices <-chan int,
-	results chan<- common.Result) error {
+	results chan<- common.Result,
+	testIDMap map[string]string) error {
 
 	var port int
 	testCaseLog := &bytes.Buffer{}
@@ -207,7 +264,20 @@ func (c *cmd) runServer(
 			CoverageData string
 			DurationMS   float32
 		}
-		postResp, err := http.Post(fmt.Sprintf("http://localhost:%v/run?%v", port, testCases[idx]), "", &bytes.Buffer{})
+		
+		var request string
+		if c.flags.mutantTracking {
+			fmt.Printf("Tracking!")
+			fileName := fmt.Sprintf("%s%d%s", "/data/dev/dredd-webgpu-testing/scripts/test_wise_mutant_tracking/track_", idx, ".txt")
+			os.Setenv("DREDD_MUTANT_TRACKING_FILE",fileName)
+			request = fmt.Sprintf("http://localhost:%v/run?%v?%v", port, testCases[idx], testIDMap[string(testCases[idx])])
+		} else {
+			fmt.Printf("No tracking!")
+			request = fmt.Sprintf("http://localhost:%v/run?%v", port, testCases[idx])
+		}
+		fmt.Printf("REQUEST: %s",request)
+
+		postResp, err := http.Post(request, "", &bytes.Buffer{})
 		if err != nil {
 			res.Error = fmt.Errorf("server POST failure. Restarting server... This can happen when there is a crash. Try running with --isolate.")
 			res.Status = common.Fail
