diff --git a/src/common/runtime/server.ts b/src/common/runtime/server.ts
index 0c13a4ff..6df3b874 100644
--- a/src/common/runtime/server.ts
+++ b/src/common/runtime/server.ts
@@ -35,6 +35,7 @@ Options:
   --force-fallback-adapter  Force a fallback adapter
   --log-to-websocket        Log to a websocket
   --u                       Flag to set on the gpu-provider as <flag>=<value>
+  --mutant-tracking         Enable test-wise mutant tracking
 
 Provides an HTTP server used for running tests via an HTTP RPC interface
 First, load some tree or subtree of tests:
@@ -79,10 +80,12 @@ if (!sys.existsSync('src/common/runtime/cmdline.ts')) {
   console.log('Must be run from repository root');
   usage(1);
 }
+
 setBaseResourcePath('out-node/resources');
 
 Colors.enabled = false;
 
+let mutantTracking = false;
 let emitCoverage = false;
 let verbose = false;
 let gpuProviderModule: GPUProviderModule | undefined = undefined;
@@ -122,6 +125,8 @@ for (let i = 0; i < sys.args.length; ++i) {
       usage(1);
     } else if (a === '--verbose') {
       verbose = true;
+    } else if (a === '--mutant-tracking' ) {
+      mutantTracking= true;
     } else {
       console.log(`unrecognized flag: ${a}`);
     }
@@ -177,9 +182,21 @@ if (verbose) {
 
   async function runTestcase(
     testcase: TestTreeLeaf,
+    testname: String,
     expectations: TestQueryWithExpectation[] = []
   ): Promise<LiveTestCaseResult> {
     const name = testcase.query.toString();
+    if (mutantTracking) {
+      if (testname != "") {
+        const mutantTrackingDir = process.env.DREDD_MUTANT_TRACKING_DIR;
+        if (mutantTrackingDir === undefined) {
+          throw new Error('DREDD_MUTANT_TRACKING_DIR must be set when mutant tracking is enabled');
+        }
+        process.env.DREDD_MUTANT_TRACKING_FILE =
+          mutantTrackingDir + '/' + testname + '.txt';
+      }
+    }
+
     const [rec, res] = log.record(name);
     await testcase.run(rec, expectations);
     return res;
@@ -192,6 +209,10 @@ if (verbose) {
         return;
       }
 
+      // Create new GPU device for each test
+      //setGPUProvider(() => gpuProviderModule!.create(gpuProviderFlags));
+
+
       const loadCasesPrefix = '/load?';
       const runPrefix = '/run?';
       const terminatePrefix = '/terminate';
@@ -211,7 +232,19 @@ if (verbose) {
           response.end(`load failed with error: ${err}\n${(err as Error).stack}`);
         }
       } else if (request.url.startsWith(runPrefix)) {
-        const name = request.url.substr(runPrefix.length);
+        // Format is runPrefix?testcase?testname for tracking
+        // Format is runPrefix?testcase for non-tracking
+        let name : string = "";
+        let testname : string = "";
+
+        if (mutantTracking) {
+          name = request.url.substring(runPrefix.length, request.url.lastIndexOf('?'));
+          testname = request.url.substring(request.url.lastIndexOf('?') + 1,);
+        }
+        else {
+          name = request.url.substring(runPrefix.length);
+        }
+
         try {
           const testcase = testcases.get(name);
           if (testcase) {
@@ -219,7 +252,7 @@ if (verbose) {
               codeCoverage.begin();
             }
             const start = performance.now();
-            const result = await runTestcase(testcase);
+            const result = await runTestcase(testcase, testname);
             const durationMS = performance.now() - start;
             const coverageData = codeCoverage !== undefined ? codeCoverage.end() : undefined;
             let message = '';
